### ISSUE TITLE
No exit status

### ISSUE THREAD

**Issue by johnmous** [s-andrews/FastQC#34]
Running fastqc for a particular file results in a `java.lang.OutOfMemoryError` but no exit status. In the case this happens inside a pipeline run, the parent process runs forever waiting for the exit status.

**Comment by s-andrews** [comment]
That could be icky.  Depending on what's happening when this happens it's possible that the parent process doesn't actually exit so the program just hangs rather than crashing.  It might be possible to put a top level trap in place to grab any uncaught exceptions and notice that the child process has gone so things don't hang.

**Comment by s-andrews** [comment]
I did some testing around this.  My concern was that we weren't transferring exit codes properly, but it looks like we are.  My test was a set of 3 files, one of which is deliberately broken:

```
bad.fastq
good.fastq
yet_more_good.fastq
```

I tested these in various ways:

```
[andrewss@headstone fastqc_bug]$ ~/FastQC/fastqc -q good.fastq
[andrewss@headstone fastqc_bug]$ echo $?
0
[andrewss@headstone fastqc_bug]$ ~/FastQC/fastqc -q bad.fastq
Failed to process bad.fastq
uk.ac.babraham.FastQC.Sequence.SequenceFormatException: ID line didn't start wit                                                         h '@'
        at uk.ac.babraham.FastQC.Sequence.FastQFile.readNext(FastQFile.java:158)
        at uk.ac.babraham.FastQC.Sequence.FastQFile.<init>(FastQFile.java:89)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:106)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:62)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.processFile(OfflineRunner.java:152)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.<init>(OfflineRunner.java:121)
        at uk.ac.babraham.FastQC.FastQCApplication.main(FastQCApplication.java:316)
[andrewss@headstone fastqc_bug]$ echo $?
1
[andrewss@headstone fastqc_bug]$ ~/FastQC/fastqc -q *fastq
Failed to process bad.fastq
uk.ac.babraham.FastQC.Sequence.SequenceFormatException: ID line didn't start with '@'
        at uk.ac.babraham.FastQC.Sequence.FastQFile.readNext(FastQFile.java:158)
        at uk.ac.babraham.FastQC.Sequence.FastQFile.<init>(FastQFile.java:89)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:106)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:62)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.processFile(OfflineRunner.java:152)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.<init>(OfflineRunner.java:121)
        at uk.ac.babraham.FastQC.FastQCApplication.main(FastQCApplication.java:316)
[andrewss@headstone fastqc_bug]$ echo $?
1
[andrewss@headstone fastqc_bug]$ ~/FastQC/fastqc -q -t 3 *fastq
Failed to process bad.fastq
uk.ac.babraham.FastQC.Sequence.SequenceFormatException: ID line didn't start with '@'
        at uk.ac.babraham.FastQC.Sequence.FastQFile.readNext(FastQFile.java:158)
        at uk.ac.babraham.FastQC.Sequence.FastQFile.<init>(FastQFile.java:89)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:106)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:62)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.processFile(OfflineRunner.java:152)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.<init>(OfflineRunner.java:121)
        at uk.ac.babraham.FastQC.FastQCApplication.main(FastQCApplication.java:316)
[andrewss@headstone fastqc_bug]$ echo $?
1

```
So all of those scenarios seem to work as expected.

**Comment by s-andrews** [comment]
I made a test file which would consume all memory and tried that.  On its own it does what you expect:

```
[andrewss@headstone fastqc_bug]~/FastQC/fastqc -q super_big.fastq
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Arrays.java:3332)
        at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
        at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:596)
        at java.lang.StringBuffer.append(StringBuffer.java:367)
        at java.io.BufferedReader.readLine(BufferedReader.java:370)
        at java.io.BufferedReader.readLine(BufferedReader.java:389)
        at uk.ac.babraham.FastQC.Sequence.FastQFile.readNext(FastQFile.java:166)
        at uk.ac.babraham.FastQC.Sequence.FastQFile.<init>(FastQFile.java:89)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:106)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:62)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.processFile(OfflineRunner.java:152)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.<init>(OfflineRunner.java:121)
        at uk.ac.babraham.FastQC.FastQCApplication.main(FastQCApplication.java:316)
[andrewss@headstone fastqc_bug]$ echo $?

```

BUT! If you mix this file in with other good files and then process them as a set then the code doesn't catch that the broken one has exited and something hangs.

```
[andrewss@headstone fastqc_bug]$ ~/FastQC/fastqc -q good.fastq super_big.fastq
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Arrays.java:3332)
        at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
        at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:596)
        at java.lang.StringBuffer.append(StringBuffer.java:367)
        at java.io.BufferedReader.readLine(BufferedReader.java:370)
        at java.io.BufferedReader.readLine(BufferedReader.java:389)
        at uk.ac.babraham.FastQC.Sequence.FastQFile.readNext(FastQFile.java:166)
        at uk.ac.babraham.FastQC.Sequence.FastQFile.<init>(FastQFile.java:89)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:106)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:62)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.processFile(OfflineRunner.java:152)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.<init>(OfflineRunner.java:121)
        at uk.ac.babraham.FastQC.FastQCApplication.main(FastQCApplication.java:316)
[Hangs here]
```

It's only a problem if there is a good sequence before the bad one.  Putting them the other way around does the right thing.

```
[andrewss@headstone fastqc_bug]$ ~/FastQC/fastqc -q super_big.fastq good.fastq
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
        at java.util.Arrays.copyOf(Arrays.java:3332)
        at java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:124)
        at java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:596)
        at java.lang.StringBuffer.append(StringBuffer.java:367)
        at java.io.BufferedReader.readLine(BufferedReader.java:370)
        at java.io.BufferedReader.readLine(BufferedReader.java:389)
        at uk.ac.babraham.FastQC.Sequence.FastQFile.readNext(FastQFile.java:166)
        at uk.ac.babraham.FastQC.Sequence.FastQFile.<init>(FastQFile.java:89)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:106)
        at uk.ac.babraham.FastQC.Sequence.SequenceFactory.getSequenceFile(SequenceFactory.java:62)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.processFile(OfflineRunner.java:152)
        at uk.ac.babraham.FastQC.Analysis.OfflineRunner.<init>(OfflineRunner.java:121)
        at uk.ac.babraham.FastQC.FastQCApplication.main(FastQCApplication.java:316)
[andrewss@headstone fastqc_bug]$ echo $?
1

```

**Comment by s-andrews** [comment]
OK, I think I've got to the bottom of this.  The OutOfMemory does get caught by our code, but because we're out of memory it seems that updating the state variables doesn't work so the program can't tell that it should be finished.

Since the behaviour of the program is somewhat undefined once we've had an out of memory error then I've changed the code so that it traps out of memory separately and reports this and then immediately exits, abandoning all attempts to process any remaining files in this session.
